lms/views/api/canvas/authorize.py

```diff
@@ -127,15 +128,31 @@ def oauth2_redirect(request):
     request_method="GET",
     route_name="canvas_api.oauth.authorize",
     renderer="lms:templates/api/oauth2/redirect_error.html.jinja2",
+    tm_active=True,
 )
 def oauth2_redirect_error(request):
     kwargs = {
         "auth_route": "canvas_api.oauth.authorize",
         "canvas_scopes": list(ALL_SCOPES),
     }
-
     if request.params.get("error") == "invalid_scope":
-        kwargs["error_code"] = request.context.js_config.ErrorCode.CANVAS_INVALID_SCOPE
+        error_code = request.context.js_config.ErrorCode.CANVAS_INVALID_SCOPE
+        kwargs["error_code"] = error_code
+
+        # Create a new DB session outside the request cycle handled by zope's
+        # transaction manager.
+        # The default one rollbacks transactions in exceptions views.
+        error_db_session = request.registry["sqlalchemy.factory"](
```

Comment:
This line is a bit verbose, get a reference to he factory and pass it a reference to the engine.

We'll probably use this in handful of other views, no need to try to make it just shorter IMO.